test/TestVectorField.cc

```diff
@@ -0,0 +1,1630 @@
+            if (response.angle_tiles_size()) {
+                // Fill PA values
+                auto tile_pa = response.angle_tiles(0);
+                GetTileData(tile_pa, downsampled_width, pa2);
+            }
+
+            // Record progress
+            progresses.push_back(response.progress());
+        };
+
+        // Do PI/PA calculations by the Frame function
+        VectorFieldCalculator vector_field_calculator(frame);
+        vector_field_calculator.DoCalculations(callback);
+
+        // Check results
+        if (file_type == CARTA::FileType::HDF5) {
+            RemoveRightAndBottomEdgeData(pi, pi2, pa, pa2, downsampled_width, downsampled_height);
+            CmpVectors(pi, pi2, 1e-5);
+            CmpVectors(pa, pa2, 1e-4);
+        } else {
+            CmpVectors(pi, pi2);
+            CmpVectors(pa, pa2);
+        }
+        CheckProgresses(progresses);
+    }
+
+    static void CalcPiAndPa(const std::string& file_path, const CARTA::FileType& file_type, int channel, int mip, bool debiasing,
+        bool fractional, double threshold, double q_error, double u_error, int& downsampled_width, int& downsampled_height,
+        std::vector<float>& pi, std::vector<float>& pa) {
+        // Create the image reader
+        std::shared_ptr<DataReader> reader = nullptr;
+        if (file_type == CARTA::FileType::HDF5) {
+            reader.reset(new Hdf5DataReader(file_path));
+        } else {
+            reader.reset(new FitsDataReader(file_path));
+        }
+
+        std::vector<float> stokes_data_i = reader->ReadXY(channel, 0);
+        std::vector<float> stokes_data_q = reader->ReadXY(channel, 1);
+        std::vector<float> stokes_data_u = reader->ReadXY(channel, 2);
+
+        // Block averaging, get downsampled data
+        int image_width = reader->Width();
+        int image_height = reader->Height();
+        downsampled_width = std::ceil((float)image_width / mip);
+        downsampled_height = std::ceil((float)image_height / mip);
+        int downsampled_area = downsampled_height * downsampled_width;
+
+        std::vector<float> downsampled_i(downsampled_area);
+        std::vector<float> downsampled_q(downsampled_area);
+        std::vector<float> downsampled_u(downsampled_area);
+
+        BlockSmooth(
+            stokes_data_i.data(), downsampled_i.data(), image_width, image_height, downsampled_width, downsampled_height, 0, 0, mip);
+        BlockSmooth(
+            stokes_data_q.data(), downsampled_q.data(), image_width, image_height, downsampled_width, downsampled_height, 0, 0, mip);
+        BlockSmooth(
+            stokes_data_u.data(), downsampled_u.data(), image_width, image_height, downsampled_width, downsampled_height, 0, 0, mip);
+
+        // Reset Q and U errors as 0 if debiasing is not used
+        if (!debiasing) {
+            q_error = u_error = 0;
+        }
+
+        // Calculate PI, FPI, and PA
+        auto calc_pi = [&](float q, float u) {
```

Comment:
A lot of this code is duplicated in the test and in the source. If the same calculations have to be used in both places, they should be factored out into common functions which both the source and the test can call.